### DISCUSSION TITLE
Views and Nested Results

### DISCUSSION THREAD

**Discussion by svarlet** [simolus3/drift#1659] [Q&A]
Hi,

I've created a view in my moor file to greatly simplify a number of queries interested in a very similar dataset. Though a number of queries previously written could benefit from the view if I knew how to deal with [Nested Results](https://drift.simonbinder.eu/docs/using-sql/drift_files/#nested-results).

Here is an example where we schedule file/folder transfers:

```sql
CREATE TABLE sync_sessions (
   id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
   global_endpoint_id INT NOT NULL UNIQUE, -- that's a GUID
   endpoint TEXT NOT NULL MAPPED BY `const UriConverter()`,
);

CREATE TABLE transfers (
    id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
    uri TEXT NOT NULL MAPPED BY `const UriConverter()`,
    status TEXT NOT NULL,
    handed_off bool NOT NULL DEFAULT false,
    sync_session INT NOT NULL,
    type ENUM(TransferType) NOT NULL,
    -- few more columns removed to keep this schema as simple as possible

    UNIQUE (sync_session, uri),

    FOREIGN KEY (sync_session)
    REFERENCES sync_sessions(id)
    ON DELETE CASCADE
    ON UPDATE CASCADE
);

CREATE TABLE successes (
    id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
    transfer INT NOT NULL UNIQUE,
    created_time INT MAPPED BY `const InstantConverter()`,
    last_modified_time INT MAPPED BY `const InstantConverter()`,
    size INT NOT NULL,
    etag TEXT,

    FOREIGN KEY (transfer)
	REFERENCES transfers (id)
	ON DELETE CASCADE
	ON UPDATE CASCADE
);

CREATE TABLE failures (
    id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
    transfer INT NOT NULL UNIQUE,
    reason TEXT NOT NULL,

    FOREIGN KEY (transfer)
	REFERENCES transfers (id)
	ON DELETE CASCADE
	ON UPDATE CASCADE
);

-- here is one of the views simplifying queries interested in the set of incomplete transfers
CREATE VIEW IF NOT EXISTS incompleteTransfers(transfer_id, endpoint_id, uri, status, handed_off)
AS
   SELECT t.id, ss.global_endpoint_id, t.uri, t.status, t.handed_off
    FROM sync_sessions AS ss
        INNER JOIN transfers as t on ss.id = t.sync_session
        LEFT JOIN successes as s on s.transfer = t.id
        LEFT JOIN failures as f on f.transfer = t.id
    WHERE s.transfer IS NULL 
        AND f.transfer IS NULL; 

-- here is a query using Drift's "nested results" feature.
-- We could improve it by querying the view (less duplicated code, more concise, clearer overall).
-- But I don't know how to maintain the current api which returns a list of `TransferWithSyncpoint`
resumables AS TransferWithSyncpoint:
    SELECT transfer.**,
        syncSession.**
    FROM transfers AS transfer
        INNER JOIN
        sync_sessions AS syncSession ON transfer.sync_session = syncSession.id
        LEFT OUTER JOIN
        successes AS sc ON sc.transfer = transfer.id
        LEFT OUTER JOIN
        failures AS f ON f.transfer = transfer.id
    WHERE 
        transfer.handed_off = 0 AND 
        transfer.status IS NOT 'NOT_STARTED' AND
        sc.transfer IS NULL AND 
        f.transfer IS NULL;
```

Here is an attempt to refactor the query above using the view #yummy :

```sql
resumable AS TransferWithSyncpoint:
    SELECT ............................
    FROM incompleteTransfers
    WHERE handed_off = 0
        AND status IS NOT 'NOT_STARTED';
```

What do I replace the dots with to get a `TransferWithSyncpoint`?

**Comment by simolus3** [comment]
At the moment we don't support `**` inside views. It's intriguing to think about it, but it would probably be very hard to support that in the generator (we also don't support `.**` in subqueries, for instance - it's just hard to think about how these should crawl up to the final result set).

I think the easiest way would just be to join from the view? Obviously there's a bit of boilerplate with the shared joins, but at least there's a bit of an improvement here.

```sql
resumable AS TransferWithSyncpoint:
    SELECT syncSession.**, transfers.**
    FROM incompleteTransfers it
      INNER JOIN sync_sessions AS syncSession ON syncSession.global_endpoint_id = it.endpoint_id
      INNER JOIN transfers AS transfer ON transfer.id = it.transfer_id     
    WHERE handed_off = 0
        AND status IS NOT 'NOT_STARTED';
```

**Reply by svarlet** [reply]
Interesting solution, and I wonder if it works. I did notice though that it would join twice (once in the view, once in this query) with sync_sessions table and twice with the transfers table. That doesn't sound terribly efficient, though in some contexts (low volumes) it may be affordable.

I wonder if there's a signal here telling me not to return all columns of a join by default, but to be specific and decisive in what columns a query is meant to return. If I followed this lead, I would not require the ** notation. Then I have the intuition that the problem discussed in this conversation is eliminated by design.

Thoughts?

**Reply by simolus3** [reply]
> I did notice though that it would join twice (once in the view, once in this query) with sync_sessions table and twice with the transfers table

It's hard to argue about query performance without seeing some numbers :) Sqlite is optimizing queries when you run them, and this includes inlining views into subqueries and considering further optimizations if possible. Sqlite may be smart enough to realize that some of the joins are identical and skip them (I haven't actually tried it).

> if there's a signal here telling me not to return all columns of a join by default, but to be specific and decisive in what columns a query is meant to return

I mean, you can just list the columns in SQL right? Just use the columns you're interested in instead of `**`. By design, `*a.*` loads all columns from `a`. You can use `**` on a subquery that only selects some columns if you want to keep nested results.
